### PR TITLE
Implement S3 storage

### PR DESCRIPTION
**Pull request by Karolk99**
## Acknowledging the stipulations set forth:
 - [ ] I hereby confirm that a Pull Request involving updates to the Software Development Kit (SDK) has been smoothly merged, currently awaits processing, or is otherwise deemed unnecessary in this context.
 - [ ] I also affirm that another Pull Request, specifically addressing updates to the documentation body (commonly referred to as 'docs'), has either been successfully incorporated, is in the process of review, or is considered superfluous under the prevailing circumstances.

This draft is not meant to be reviewed entirely. The purpose of this PR is to show one of two possible solutions that can be fully implemented.
In this case, HLS is sent to S3 by `Membrane.HTTPAdaptiveStream.Storage` implementation so every segment and manifest iteration is sent right after creation.
Apart from sending it to the S3 bucket each segment is saved on the Jellyfish server.
That allows us to serve live HLS simultaneously and fix some problems in postprocessing that may occur during the stream.

The big disadvantage is that we cannot push HLS to S3 and simultaneously serve LL-HLS (it is technically possible but It would require a lot of work).

### CODE REVIEW

**Comment by Karolk99** [comment]
Send the whole stream at once on the Engine termination

In this scenario, we would spawn (if needed) the process responsible for sending the stream to S3 on the `:add_component` event.
This process would monitor the Engine process and on the Engine termination, It would do the sending.
This solution requires some constraints, ex: the stream cannot be removed before the end of postprocessing, so the option `persist` has to be reconsidered.
On the other hand, It gives us a lot of flexibility with the stream, ex: we can validate and repair the stream if needed before sending it to S3,  we can use LL-HLS ( it's important, we can't do it easily in the storage scenario)

```
@impl true
  def handle_call({:add_component, component_type, options}, _from, state) do
    options =
      Map.merge(
        %{engine_pid: state.engine_pid, room_id: state.id},
        options
      )

    with :ok <- check_component_allowed(component_type, state),
         {:ok, component} <- Component.new(component_type, options) do
      state = put_in(state, [:components, component.id], component)
      if component_type == HLS do
         on_hls_startup(state.id, component.metadata)

         # spawning process responsible for s3
         maybe_init_s3_process(state.engine_pid, s3_credentials)
     end
      :ok = Engine.add_endpoint(state.engine_pid, component.engine_endpoint, id: component.id)
     .
     .
     .
  end
```

**Review by Rados13 — COMMENTED** [review]
In the end, I would prefer the second approach. Because we could use LL-HLS for streaming with putting to S3, I think if we don't have specific requirements that we have to do something in realtime we should put as much as possible into postprocessing, because it would be safer and easier for testing, and debugging.

File: lib/jellyfish/component/hls/storage/s3.ex [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    bucket
+    |> ExAws.S3.put_object(filepath, content, opts)
+    |> ExAws.request(client)
+  end
+
+  defp remove_from_s3(filepath, %{bucket: bucket, client: client}) do
+    bucket
+    |> ExAws.S3.delete_object(filepath)
+    |> ExAws.request(client)
+  end
```
[the remark this answers is not in the document]
Comment (Karolk99) [reply lib/jellyfish/component/hls/storage/s3.ex:86]: If it's meant to be removed that means this file is garbage so I don't see a point in keeping it on S3

File: lib/jellyfish/component/hls/storage/s3.ex [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+          raise "The S3 storage doesn't support ll-hls."
+
+        %{mode: :binary, type: :header} ->
+          write_to_s3(s3_path, content, state, "application/vnd.apple.mpegurl")
+          write_to_file(directory, name, content, [:binary])
+
+        %{mode: :text, type: :manifest} ->
+          write_to_s3(s3_path, content, state)
+          write_to_file(directory, name, content)
+      end
```
[the remark this answers is not in the document]
Comment (Karolk99) [reply lib/jellyfish/component/hls/storage/s3.ex:49]: As you can see we will only raise if writing to file won't succeed. We can always give up on saving to file but we won't have a backup that way.

**Review by mickel8 — COMMENTED** [review]
In the second approach, did we consider starting uploading as soon as possible (before we close a room)?

The only thing I would like to achieve is to be able to upload to s3 using RTC Engine :thinking:

**Pull request closed without merging**